File: Taverne/src/app/model/inventaire/boisson.ts
import { Bar } from './bar';
import { Utilisation } from './utilisation';

export class Boisson {
  constructor(
    protected _type?: string,
    protected _id?: number,
    protected _bar?: Bar,
    protected _nom?: string,
    protected _prixHT?: number,
    protected _prixHThh?: number,
    protected _tva?: number,
    protected _utilisations?: Utilisation[],
    protected _description?: string,
    protected _urlImage?: string
  ) {}

  get utilisations(): Utilisation[] | undefined {
    return this._utilisations;
  }

  set utilisations(value: Utilisation[] | undefined) {
    this._utilisations = value;
  }

  get type(): string | undefined {
    return this._type;
  }

  get id(): number | undefined {
    return this._id;
  }

  get bar(): Bar | undefined {
    return this._bar;
  }

  get nom(): string | undefined {
    return this._nom;
  }

  get prixHT(): number | undefined {
    return this._prixHT;
  }

  get prixHThh(): number | undefined {
    return this._prixHThh;
  }

  get tva(): number | undefined {
    return this._tva;
  }

  set type(value: string | undefined) {
    this._type = value;
  }

  set id(value: number | undefined) {
    this._id = value;
  }

  set bar(value: Bar | undefined) {
    this._bar = value;
  }

  set nom(value: string | undefined) {
    this._nom = value;
  }

  set prixHT(value: number | undefined) {
    this._prixHT = value;
  }

  set prixHThh(value: number | undefined) {
    this._prixHThh = value;
  }

  set tva(value: number | undefined) {
    this._tva = value;
  }
  /**
   * Getter $description
   * @return {string}
   */
  public get description(): string | undefined {
    return this._description;
  }

  /**
   * Setter $description
   * @param {string} value
   */
  public set description(value: string | undefined) {
    this._description = value;
  }

  /**
   * Getter $urlImage
   * @return {string}
   */
  public get urlImage(): string | undefined {
    return this._urlImage;
  }

  /**
   * Setter $urlImage
   * @param {string} value
   */
  public set urlImage(value: string | undefined) {
    this._urlImage = value;
  }
}

File: TaverneSpringBoot/src/main/java/soprajc/TaverneSpringBoot/model/inventaire/Boisson.java
package soprajc.TaverneSpringBoot.model.inventaire;

import java.util.Objects;
import java.util.Set;

import javax.persistence.CascadeType;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Version;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonView;

import soprajc.TaverneSpringBoot.model.JsonViews;

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "Type_Boisson")
@JsonTypeInfo(
		  use = JsonTypeInfo.Id.NAME, 
		  include = JsonTypeInfo.As.PROPERTY, 
		  property = "type",
		  visible = true)
		@JsonSubTypes({ 
		  @Type(value = Alcool.class, name = "alcool"), 
		  @Type(value = Soft.class, name = "soft") 
		})
public abstract class Boisson {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)

	@JsonView({ JsonViews.Common.class, JsonViews.BoissonByBar.class })
	protected Long id;
	@JsonView({ JsonViews.Common.class, JsonViews.BoissonByBar.class })
	protected String nom;
	@JsonView({ JsonViews.Common.class, JsonViews.BoissonByBar.class })
	protected double prixHT;
	@JsonView({ JsonViews.Common.class, JsonViews.BoissonByBar.class })
	protected double prixHThh;
	@JsonView({ JsonViews.Common.class, JsonViews.BoissonByBar.class })
	protected double tva;

	@JsonView(JsonViews.Common.class)
	@ManyToOne
	@JoinColumn(name = "id_bar")
	protected Bar bar;

	@OneToMany(mappedBy = "boisson", cascade = CascadeType.ALL)
	@JsonView(JsonViews.Common.class)
	protected Set<Utilisation> utilisations;

	public Boisson() {
	}

	public Boisson(String nom, double prixHT, double prixHThh, double tva, Bar bar, Set<Utilisation> utilisations) {
		this.nom = nom;
		this.prixHT = prixHT;
		this.prixHThh = prixHThh;
		this.tva = tva;
		this.bar = bar;
		this.utilisations = utilisations;
	}

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public String getNom() {
		return nom;
	}

	public void setNom(String nom) {
		this.nom = nom;
	}

	public double getPrixHT() {
		return prixHT;
	}

	public void setPrixHT(double prixHT) {
		this.prixHT = prixHT;
	}

	public double getPrixHThh() {
		return prixHThh;
	}

	public void setPrixHThh(double prixHThh) {
		this.prixHThh = prixHThh;
	}

	public double getTva() {
		return tva;
	}

	public void setTva(double tva) {
		this.tva = tva;
	}

	public Bar getBar() {
		return bar;
	}

	public void setBar(Bar bar) {
		this.bar = bar;
	}

	public Set<Utilisation> getUtilisations() {
		return utilisations;
	}

	public void setUtilisations(Set<Utilisation> utilisations) {
		this.utilisations = utilisations;
	}

	@Version
	private int version;

	@Override
	public int hashCode() {
		return Objects.hash(bar, id, nom, prixHT, prixHThh, tva, utilisations);
	}

	public int getVersion() {
		return version;
	}

	public void setVersion(int version) {
		this.version = version;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Boisson other = (Boisson) obj;
		return Objects.equals(bar, other.bar) && Objects.equals(id, other.id) && Objects.equals(nom, other.nom)
				&& Double.doubleToLongBits(prixHT) == Double.doubleToLongBits(other.prixHT)
				&& Double.doubleToLongBits(prixHThh) == Double.doubleToLongBits(other.prixHThh)
				&& Double.doubleToLongBits(tva) == Double.doubleToLongBits(other.tva)
				&& Objects.equals(utilisations, other.utilisations);
	}

}
